example/ipsec_offload/odp_ipsec_offload.c

```diff
@@ -0,0 +1,855 @@
+ *
+ * @return PKT_CONTINUE if good, supported packet else PKT_DROP
+ */
+static pkt_disposition_e do_input_verify(odp_packet_t pkt)
+{
+	if (odp_unlikely(odp_packet_has_error(pkt))) {
+		odp_packet_free(pkt);
+		return PKT_DROP;
+	}
+
+	if (!odp_packet_has_eth(pkt)) {
+		odp_packet_free(pkt);
+		return PKT_DROP;
+	}
+
+	if (!odp_packet_has_ipv4(pkt)) {
+		odp_packet_free(pkt);
+		return PKT_DROP;
+	}
+
+	return PKT_CONTINUE;
+}
+
+/**
+ * Packet Processing - Route lookup in forwarding database
+ *
+ * @param pkt  Packet to route
+ *
+ * @return PKT_CONTINUE if route found else PKT_DROP
+ */
+static
+pkt_disposition_e do_route_fwd_db(odp_packet_t pkt)
+{
+	odph_ipv4hdr_t *ip = (odph_ipv4hdr_t *)odp_packet_l3_ptr(pkt, NULL);
+	fwd_db_entry_t *fwd_entry;
+	ipsec_cache_entry_t *ipsec_entry;
+	uint32_t	sip, dip;
+	uint64_t	hash;
+	odp_flow_entry_t *flow = NULL;
+
+	if (ip->ttl > 1) {
+		ip->ttl -= 1;
+		if (ip->chksum >= odp_cpu_to_be_16(0xffff - 0x100))
+			ip->chksum += odp_cpu_to_be_16(0x100) + 1;
+		else
+			ip->chksum += odp_cpu_to_be_16(0x100);
+	} else {
+		odp_packet_free(pkt);
+		return PKT_DROP;
+	}
+
+	sip = odp_be_to_cpu_32(ip->src_addr);
+	dip = odp_be_to_cpu_32(ip->dst_addr);
+
+	hash = calculate_flow_hash(sip, dip);
+
+	flow = odp_route_flow_lookup_in_bucket(sip, dip,
+					       &flow_table[hash &
+							   (bucket_count - 1)]);
+	if (flow) {
+		odp_ipsec_out_param_t params;
+
+do_opt:
+		odp_packet_user_ptr_set(pkt, &flow->out_port);
+		if (flow->out_port.sa == ODP_IPSEC_SA_INVALID)
+			return PKT_CONTINUE;
+
+		/* Initialize parameters block */
+		params.sa = &flow->out_port.sa;
+		params.opt = NULL;
+		params.num_sa = 1;
+		params.num_opt = 1;
+
+		/* Issue ipsec request */
+		if (odp_unlikely(odp_ipsec_out_enq(&pkt, 1, &params) < 0)) {
+			EXAMPLE_DBG("Unable to out enqueue\n");
+			odp_packet_free(pkt);
+			return PKT_DROP;
+		}
+		return PKT_POSTED;
+	} else {
+		/*Check into Routing table*/
+		fwd_entry = find_fwd_db_entry(dip);
+		if (fwd_entry) {
+			/*Entry found. Updated in Flow table first.*/
+			flow = calloc(1, sizeof(odp_flow_entry_t));
+			if (!flow) {
+				EXAMPLE_ABORT("Failure to allocate memory");
+				return PKT_DROP;
+			}
+			flow->l3_src = sip;
+			flow->l3_dst = dip;
+			flow->out_port.pktout = fwd_entry->pktout;
+			memcpy(flow->out_port.addr.addr,
+			       fwd_entry->src_mac,
+			       ODPH_ETHADDR_LEN);
+			memcpy(flow->out_port.next_hop_addr.addr,
+			       fwd_entry->dst_mac,
+			       ODPH_ETHADDR_LEN);
+			ipsec_entry = find_ipsec_cache_entry_out(sip, dip);
+			if (ipsec_entry)
+				flow->out_port.sa = ipsec_entry->sa;
+			else
+				flow->out_port.sa = ODP_IPSEC_SA_INVALID;
+			flow->next = NULL;
+			/*Insert new flow into flow cache table*/
+			odp_route_flow_insert_in_bucket(flow, &flow_table[hash &
+							(bucket_count - 1)]);
+			goto do_opt;
+		} else {
+			EXAMPLE_DBG("No flow match found. Packet is dropped.\n");
+			odp_packet_free(pkt);
+			return PKT_DROP;
+		}
+	}
+}
```

Comment:
Checkpatch doesn't like this and the `goto` is a bit ugly. Might a restructure along the following lines be simpler and cleaner?
```
if (!flow) {
        /*Check into Routing table*/
        fwd_entry = find_fwd_db_entry(dip);
     
        if (!fwd_entry) {
			EXAMPLE_DBG("No flow match found. Packet is dropped.\n");
			odp_packet_free(pkt);
 	                return PKT_DROP;
        }
        ... flow setup processing, fall though without needing a goto.
}

...flow processing goes here
```